src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs

```diff
@@ -108,14 +108,33 @@ public AuthorizeFilter(string policy)
 
         bool IFilterFactory.IsReusable => true;
 
+        // Computes the actual policy for this filter using either Policy or PolicyProvider + AuthorizeData
+        private Task<AuthorizationPolicy> ComputePolicyAsync()
+        {
+            if (Policy != null)
+            {
+                return Task.FromResult(Policy);
+            }
+            if (PolicyProvider == null)
+            {
+                throw new InvalidOperationException(
+                    Resources.FormatAuthorizeFilter_AuthorizationPolicyCannotBeCreated(
+                        nameof(AuthorizationPolicy),
+                        nameof(IAuthorizationPolicyProvider)));
+            }
+
+            return AuthorizationPolicy.CombineAsync(PolicyProvider, AuthorizeData);
+        }
+
         private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFilterContext context)
         {
             if (_effectivePolicy != null)
```

Comment:
Thinking more, with 'combining of authorization policies' support (which was the original intent too), `AuthorizeFilter` doesn't anymore work like how filters are supposed to. It just feels like a 'stage' (like model binding) in an action execution pipeline where a single thing (you can either call it a 'closest authorizefilter' as we are currently doing or 'authorizationmanager' or something) is just gathering metadata from global, controller and action and combining them. We might as well can remove Authorization to be a filter in future and let it be a stage instead. This is of course a breaking change and cannot be done until 3.0, if we even consider it.

Reply:
Yes this is really complicated code, would be much simpler if it could just a single loop instead of this filter/combining/closest match business...